File: central-assistant-py/spike/masking.py
"""
Python port of central-assistant/masking.mjs (Stage 5) — same contract:

  OUTBOUND: every string that leaves for the LLM passes mask_text. Real vessel
  names, person names, IMO numbers and DB UUIDs become stable per-request tokens
  [VESSEL_1], [PERSON_1], [IMO_1], [ID_1].
  INBOUND: the final answer passes unmask_text; LLM-produced tool arguments pass
  unmask_json BEFORE a module runs them (modules always work on real data).

Failure policy is unchanged: mask failure -> raise (request refused before any
LLM call); unmapped token on the way back -> left VISIBLE and recorded in
`warnings`. The '__MASK_FAIL_TEST__' seam is kept so the fail-closed path stays
provable end to end.
"""
from __future__ import annotations

import re
from typing import Any

NAME_KEY_RE = re.compile(
    r"(vessel_?name|user_?name|full_?name|display_?name|_by_name|_by$|approver|observer|"
    r"reported_by|completed_by|created_by|updated_by|master|chief|requestedBy|assignee)",
    re.I,
)
UUID_RE = re.compile(r"\b[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\b", re.I)
IMO_RE = re.compile(r"\b\d{7}\b")
# Accepts the token WITH or WITHOUT its brackets: models (gpt-4o-mini, spike 11-Sep) sometimes
# emit "VESSEL_1" / "**VESSEL_1**" — markdown eats the brackets — in answers AND in tool
# arguments. A bare VESSEL_1 never occurs in real text, so the looser match is safe.
TOKEN_RE = re.compile(r"\[?\b(VESSEL|PERSON|IMO|ID)_(\d+)\b\]?")

VESSEL_SIGNAL = re.compile(r"^(imo|imo_?number|vessel_?id|vuuid|v_?code|vessel_?code|flag)$", re.I)
PERSON_SIGNAL = re.compile(r"^(rank|rank_?name|email|crew_?id|employee_?id|designation|department)$", re.I)
BARE_NAME_KEY = re.compile(r"^(name|full_?name|title)$", re.I)
IMO_KEY = re.compile(r"^(imo|imo_?number)$", re.I)


class Masker:
    def __init__(self) -> None:
        self._real_to_token: dict[str, str] = {}   # lowercased real -> token
        self._token_to_real: dict[str, str] = {}   # token -> real (original casing)
        self._counters = {"VESSEL": 0, "PERSON": 0, "IMO": 0, "ID": 0}
        self._name_re: re.Pattern[str] | None = None
        self.warnings: list[str] = []

    # -- registration -------------------------------------------------------
    def register(self, value: Any, kind: str) -> str | None:
        v = str(value if value is not None else "").strip()
        if len(v) < 3:
            return None
        key = v.lower()
        if key in self._real_to_token:
            return self._real_to_token[key]
        self._counters[kind] += 1
        token = f"[{kind}_{self._counters[kind]}]"
        self._real_to_token[key] = token
        self._token_to_real[token] = v
        self._name_re = None  # rebuild lazily
        return token

    def _build_name_re(self) -> re.Pattern[str] | None:
        names = [k for k in self._real_to_token if not UUID_RE.fullmatch(k) and not re.fullmatch(r"\d{7}", k)]
        if not names:
            return None
        names.sort(key=len, reverse=True)
        return re.compile(r"(?<!\w)(" + "|".join(re.escape(n) for n in names) + r")(?!\w)", re.I)

    def learn_from_json(self, obj: Any) -> None:
        """Learn identifier values from a data-tool result (walks the JSON, sibling-aware)."""
        if obj is None:
            return
        if isinstance(obj, list):
            for item in obj:
                self.learn_from_json(item)
            return
        if isinstance(obj, dict):
            keys = list(obj.keys())
            has_vessel = any(VESSEL_SIGNAL.match(str(k)) for k in keys)
            has_person = any(PERSON_SIGNAL.match(str(k)) for k in keys)
            for k, v in obj.items():
                k = str(k)
                if isinstance(v, str):
                    if NAME_KEY_RE.search(k):
                        self.register(v, "VESSEL" if re.search("vessel", k, re.I) else "PERSON")
                    elif BARE_NAME_KEY.match(k) and (has_vessel or has_person):
                        self.register(v, "VESSEL" if has_vessel else "PERSON")
                    elif IMO_KEY.match(k) and re.fullmatch(r"\d{6,8}", v.strip()):
                        self.register(v.strip(), "IMO")
                elif isinstance(v, (int, float)) and not isinstance(v, bool) and IMO_KEY.match(k):
                    self.register(str(v), "IMO")
                self.learn_from_json(v)

    # -- outbound -----------------------------------------------------------
    def mask_text(self, s: Any) -> Any:
        if not isinstance(s, str) or not s:
            return s
        if "__MASK_FAIL_TEST__" in s:
            raise RuntimeError("masking self-test failure requested")
        out = s
        if self._name_re is None:
            self._name_re = self._build_name_re()
        if self._name_re is not None:
            out = self._name_re.sub(lambda m: self._real_to_token.get(m.group(0).lower(), m.group(0)), out)
        out = UUID_RE.sub(lambda m: self._real_to_token.get(m.group(0).lower()) or self.register(m.group(0), "ID") or m.group(0), out)
        out = IMO_RE.sub(lambda m: self._real_to_token.get(m.group(0)) or self.register(m.group(0), "IMO") or m.group(0), out)
        return out

    def mask_json(self, obj: Any) -> Any:
        """Deep-mask every string in a JSON-like structure (returns a copy)."""
        if isinstance(obj, str):
            return self.mask_text(obj)
        if isinstance(obj, list):
            return [self.mask_json(x) for x in obj]
        if isinstance(obj, dict):
            return {k: self.mask_json(v) for k, v in obj.items()}
        return obj

    # -- inbound ------------------------------------------------------------
    def unmask_text(self, s: Any) -> Any:
        if not isinstance(s, str) or not s:
            return s

        def repl(m: re.Match[str]) -> str:
            tok = f"[{m.group(1)}_{m.group(2)}]"  # canonical form, brackets or not
            real = self._token_to_real.get(tok)
            if real is None:
                self.warnings.append(f"unmapped token left visible: {tok}")
                return tok  # visible placeholder beats a guess or a leak
            return real

        return TOKEN_RE.sub(repl, s)

    def unmask_json(self, obj: Any) -> Any:
        """Restore real values inside LLM-produced tool arguments (deep)."""
        if isinstance(obj, str):
            return self.unmask_text(obj)
        if isinstance(obj, list):
            return [self.unmask_json(x) for x in obj]
        if isinstance(obj, dict):
            return {k: self.unmask_json(v) for k, v in obj.items()}
        return obj

    def size(self) -> int:
        return len(self._token_to_real)

File: central-assistant/masking.mjs
/**
 * Stage 5 — mask-out / un-mask-in (enterprise plan B3; implementation plan Stage 5).
 *
 * PRINCIPLE — one choke point per direction:
 *   OUTBOUND (to OpenAI): every string that leaves for the LLM/embeddings passes
 *   maskText/maskMessages. Real vessel names, person names, IMO numbers and DB
 *   UUIDs are replaced by stable per-request tokens: [VESSEL_1], [PERSON_1],
 *   [IMO_1], [ID_1].
 *   INBOUND (from OpenAI): the final answer passes unmaskText — tokens become the
 *   real values again, so the user reads real names and OpenAI never saw them.
 *   Tool-call arguments produced by the LLM contain tokens; unmaskJson restores
 *   them BEFORE the module Data API is called (modules always work on real data).
 *
 * WHAT is masked:
 *   - names registered up front (context.vesselName, the caller's userName)
 *   - names LEARNED from data-tool results: values of name-bearing fields
 *     (vessel_name, *_by, *_by_name, approver, observer, master, full_name, …)
 *   - UUIDs (DB identifiers) and standalone 7-digit numbers (IMO shape) found in
 *     any outbound string — masked on sight, mapped for restore.
 *   Name replacement is WHOLE-NAME, word-boundary, case-insensitive — a vessel
 *   called "Gas Mia" masks only the full name, never the word "gas"; a string
 *   that merely LOOKS like a vessel name but is not ours is untouched (we mask
 *   OUR identifiers, not the concept of names).
 *
 * FAILURE POLICY (owner-confirmed direction):
 *   - Masking failure  → the REQUEST FAILS before any LLM call (never send raw).
 *   - Un-mask failure (unknown token in the answer) → the placeholder stays
 *     VISIBLE and a warning is logged. A visible [VESSEL_9] is honest and
 *     debuggable; silently guessing or leaking is not.
 *   Test seam: a message containing '__MASK_FAIL_TEST__' makes maskText throw,
 *   so the fail-closed path is provable end-to-end.
 */

const NAME_KEY_RE = /(vessel_?name|user_?name|full_?name|display_?name|_by_name|_by$|approver|observer|reported_by|completed_by|created_by|updated_by|master|chief|requestedBy|assignee)/i;
const UUID_RE = /\b[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\b/gi;
const IMO_RE = /\b\d{7}\b/g;
const TOKEN_RE = /\[(VESSEL|PERSON|IMO|ID)_(\d+)\]/g;

const escapeRe = (s) => s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

export function createMasker() {
  const realToToken = new Map(); // lowercased real -> token
  const tokenToReal = new Map(); // token -> real (original casing)
  const counters = { VESSEL: 0, PERSON: 0, IMO: 0, ID: 0 };
  let nameRe = null; // combined regex over registered names, longest-first
  const warnings = [];

  function register(value, kind) {
    const v = String(value ?? '').trim();
    if (v.length < 3) return null;
    const key = v.toLowerCase();
    if (realToToken.has(key)) return realToToken.get(key);
    const token = `[${kind}_${++counters[kind]}]`;
    realToToken.set(key, token);
    tokenToReal.set(token, v);
    nameRe = null; // rebuild lazily
    return token;
  }

  function buildNameRe() {
    const names = [...realToToken.keys()]
      .filter((k) => !UUID_RE.test(k) && !/^\d{7}$/.test(k))
      .sort((a, b) => b.length - a.length);
    UUID_RE.lastIndex = 0;
    if (!names.length) return null;
    return new RegExp(`(?<![\\w])(${names.map(escapeRe).join('|')})(?![\\w])`, 'gi');
  }

  // A bare `name`/`title` field is a real identifier only when its OBJECT also
  // carries a signal of what it names — a vessel (imo/code/vessel id) or a person
  // (rank/email/crew). This is what stops manual excerpts like {title:"Save as
  // Draft"} being masked while {name:"Gas Mia", imoNumber:...} IS masked.
  const VESSEL_SIGNAL = /^(imo|imo_?number|vessel_?id|vuuid|v_?code|vessel_?code|flag)$/i;
  const PERSON_SIGNAL = /^(rank|rank_?name|email|crew_?id|employee_?id|designation|department)$/i;
  const BARE_NAME_KEY = /^(name|full_?name|title)$/i;

  /** Learn identifier values from a data-tool result (walks the JSON, sibling-aware). */
  function learnFromJson(obj) {
    const walk = (o) => {
      if (o === null || o === undefined) return;
      if (Array.isArray(o)) return o.forEach(walk);
      if (typeof o === 'object') {
        const keys = Object.keys(o);
        const hasVesselSignal = keys.some((k) => VESSEL_SIGNAL.test(k));
        const hasPersonSignal = keys.some((k) => PERSON_SIGNAL.test(k));
        for (const [k, v] of Object.entries(o)) {
          if (typeof v === 'string') {
            if (NAME_KEY_RE.test(k)) {
              register(v, /vessel/i.test(k) ? 'VESSEL' : 'PERSON');
            } else if (BARE_NAME_KEY.test(k) && (hasVesselSignal || hasPersonSignal)) {
              register(v, hasVesselSignal ? 'VESSEL' : 'PERSON');
            } else if (/^(imo|imo_?number)$/i.test(k) && /^\d{6,8}$/.test(v.trim())) {
              register(v.trim(), 'IMO');
            }
          } else if (typeof v === 'number' && /^(imo|imo_?number)$/i.test(k)) {
            register(String(v), 'IMO');
          }
          walk(v);
        }
      }
    };
    walk(obj);
  }

  function maskText(s) {
    if (typeof s !== 'string' || !s) return s;
    if (s.includes('__MASK_FAIL_TEST__')) throw new Error('masking self-test failure requested');
    let out = s;
    if (!nameRe) nameRe = buildNameRe();
    if (nameRe) {
      nameRe.lastIndex = 0;
      out = out.replace(nameRe, (m) => realToToken.get(m.toLowerCase()) || m);
    }
    out = out.replace(UUID_RE, (m) => realToToken.get(m.toLowerCase()) || register(m, 'ID'));
    out = out.replace(IMO_RE, (m) => realToToken.get(m) || register(m, 'IMO'));
    return out;
  }

  /** Deep-mask every string in an OpenAI messages array (returns a copy). */
  function maskMessages(messages) {
    return messages.map((m) => {
      const c = { ...m };
      if (typeof c.content === 'string') c.content = maskText(c.content);
      return c;
    });
  }

  function unmaskText(s) {
    if (typeof s !== 'string' || !s) return s;
    TOKEN_RE.lastIndex = 0;
    return s.replace(TOKEN_RE, (tok) => {
      const real = tokenToReal.get(tok);
      if (real === undefined) {
        warnings.push(`unmapped token left visible: ${tok}`);
        return tok; // visible placeholder beats a guess or a leak
      }
      return real;
    });
  }

  /** Restore real values inside LLM-produced tool arguments (deep). */
  function unmaskJson(obj) {
    if (typeof obj === 'string') return unmaskText(obj);
    if (Array.isArray(obj)) return obj.map(unmaskJson);
    if (obj && typeof obj === 'object') {
      const out = {};
      for (const [k, v] of Object.entries(obj)) out[k] = unmaskJson(v);
      return out;
    }
    return obj;
  }

  return {
    register, learnFromJson, maskText, maskMessages, unmaskText, unmaskJson,
    warnings, size: () => tokenToReal.size,
  };
}
